File: plugins/extbench/fixtures/goja-js/validate.js
// validate(payload) — normalize email + check shape.
// Payload shape: { email: string, age: number }
// Result shape: { ok: bool, email?: string, age?: number, error?: string }
//
// Goja exposes this as globalThis.validate after the script runs once.
function validate(input) {
  if (input === null || typeof input !== "object") {
    return { ok: false, error: "input must be an object" };
  }
  var email = input.email;
  var age = input.age;
  if (typeof email !== "string" || email.length === 0) {
    return { ok: false, error: "email required" };
  }
  if (typeof age !== "number" || age < 0 || age > 150) {
    return { ok: false, error: "age out of range" };
  }
  var normalized = email.trim().toLowerCase();
  // Basic shape check: one '@' with non-empty local + domain containing a dot.
  var at = normalized.indexOf("@");
  if (at <= 0 || at === normalized.length - 1) {
    return { ok: false, error: "email shape" };
  }
  var domain = normalized.substring(at + 1);
  if (domain.indexOf(".") < 0) {
    return { ok: false, error: "email domain" };
  }
  return { ok: true, email: normalized, age: age };
}

File: plugins/extbench/fixtures/wazero-cpython-wasi/validate.py
"""validate(email, age) -- normalize email + check shape.

Same semantics as the goja / wazero-as / pyvm / starlark / native fixtures.

In a WASI-command-module deployment of CPython, this script reads JSON
from stdin and writes JSON to stdout. The host shim (see README) wraps
that into our pointer-based ABI.
"""

import json
import sys


def validate(payload):
    if not isinstance(payload, dict):
        return {"ok": False, "error": "input must be an object"}
    email = payload.get("email")
    age = payload.get("age")
    if not isinstance(email, str) or len(email) == 0:
        return {"ok": False, "error": "email required"}
    if not isinstance(age, int) or age < 0 or age > 150:
        return {"ok": False, "error": "age out of range"}
    normalized = email.strip().lower()
    at = normalized.find("@")
    if at <= 0 or at == len(normalized) - 1:
        return {"ok": False, "error": "email shape"}
    domain = normalized[at + 1 :]
    if domain.find(".") < 0:
        return {"ok": False, "error": "email domain"}
    return {"ok": True, "email": normalized, "age": age}


if __name__ == "__main__":
    raw = sys.stdin.read()
    payload = json.loads(raw)
    result = validate(payload)
    sys.stdout.write(json.dumps(result))
